internal/dataplane/parser/parser.go

```diff
@@ -99,7 +100,10 @@ func (p *Parser) Build() (*kongstate.KongState, error) {
 	result.FillPlugins(p.logger, p.storer)
 
 	// generate Certificates and SNIs
-	result.Certificates = getCerts(p.logger, p.storer, ingressRules.SecretNameToSNIs)
+	ingressCerts := getCerts(p.logger, p.storer, ingressRules.SecretNameToSNIs)
+	gatewayCerts := getGatewayCerts(p.logger, p.storer)
+	// note that ingress-derived certificates will take precedence over gateway-derived certificates for SNI assignment
+	result.Certificates = mergeCerts(p.logger, ingressCerts, gatewayCerts)
```

Comment:
It's basically if you have a TLS Ingress with `host: collision.example.com` and then create an HTTPS Listener with `hostname: collision.example.com`, the certificate on the Ingress wins. This ordering is basically arbitrary--someone has to win, but there's no rule that actually says who that is. If anything, Ingress is the established, stable implementation, so it's the one that wins, but that's a weak justification.

I think in practice I'd just recommend not using a mixture of Gateway and Ingress if possible, though that doesn't really save you since each can cause collisions on its own. Ingress has no protection against using the same host in different Ingresses (and we already did this sort of collision resolution, albeit silently for some reason) and Gateway only protects against it if you don't collapse Gateway resources (we do).